lib/Dialect/Poly/IR/Polynomial.cpp

```diff
@@ -0,0 +1,73 @@
+#include "include/Dialect/Poly/IR/Polynomial.h"
+
+#include "include/Dialect/Poly/IR/PolynomialDetail.h"
+#include "llvm/ADT/APInt.h"            // from @llvm-project
+#include "llvm/ADT/SmallString.h"      // from @llvm-project
+#include "llvm/Support/raw_ostream.h"  // from @llvm-project
+#include "mlir/IR/MLIRContext.h"       // from @llvm-project
+
+namespace mlir {
+namespace heir {
+namespace poly {
+
+MLIRContext *Polynomial::getContext() const { return terms->context; }
+
+ArrayRef<Monomial> Polynomial::getTerms() const { return terms->terms(); }
+
+Polynomial Polynomial::fromMonomials(ArrayRef<Monomial> monomials,
+                                     MLIRContext *context) {
+  auto assignCtx = [context](detail::PolynomialStorage *storage) {
+    storage->context = context;
+  };
+
+  // A polynomial's terms are canonically stored in order of increasing degree.
+  llvm::OwningArrayRef<Monomial> monomials_copy =
+      llvm::OwningArrayRef<Monomial>(monomials);
+  std::sort(monomials_copy.begin(), monomials_copy.end());
+
+  StorageUniquer &uniquer = context->getAttributeUniquer();
```

Comment:
not sure if there is a good answer here but what is the difference / purpose of the attribute uniquer versus any other type?

Reply:
So I don't know why the MLIRContext has multiple uniquers (they're all the same type), but I figure they might be for conflict avoidance of some sort? Like if you want separate instances of a type de-duped for attributes vs types vs affine maps? Not really sure.